### PR TITLE
Added new flag `--require-https`

### LINKED ISSUE #192: Feature request: Check usage of HTTP when HTTPS is available
Often time I'd prefer to use HTTPS urls when it's available, and gets an warning if I didn't use HTTP instead.

It seems that there's no such functionality currently, and it'd be great if users can opt-in for this feature.

### PR DESCRIPTION
**Pull request by lebensterben**
Added new CLI/Config File flag `--require-https`.
When it's presented, `lychee` additionally checks whether HTTPS encryption is
available for HTTP URLs. If so, it's treated as errors.

Closes #192

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,8 +3,8 @@
 ![Rust](https://github.com/hello-rust/lychee/workflows/Rust/badge.svg)
 [![docs.rs](https://docs.rs/lychee/badge.svg)](https://docs.rs/lychee)
 
-⚡ A fast, async, resource-friendly link checker written in Rust.  
-Finds broken hyperlinks and mail addresses inside Markdown, HTML, reStructuredText, or any other text file or website!  
+⚡ A fast, async, resource-friendly link checker written in Rust.
```
Comment (mre) [inline README.md:6]: Can we add back the two spaces at the end? They are significant because they result in a newline, which is the formatting I prefer in that case.

Comment (lebensterben) [reply README.md:6]: They're done by my formatter...

File: src/client.rs [hunk 2]
```diff
@@ -172,7 +176,23 @@ impl Client {
             return Ok(Response::new(request.uri, Status::Excluded, request.source));
         }
         let status = match request.uri {
-            Uri::Website(ref url) => self.check_website(&url).await,
+            Uri::Website(ref url) => match self.check_website(url).await {
+                Status::Ok(code) if self.prefer_https && url.scheme() == "http" => {
+                    let mut https_url = url.clone();
```
Comment (mre) [inline src/client.rs:181]: Can we move that block into a function; e.g. `is_https_available(&url)`?

File: src/client.rs [hunk 3]
```diff
@@ -172,7 +176,23 @@ impl Client {
...
+                    https_url
+                        .set_scheme("https")
+                        .unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() });
```
Comment (mre) [inline src/client.rs:184]: Not a fan of the `.unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() })` part. I don't think it can fail, but can't we just use `?` instead?

File: src/client.rs [hunk 4]
```diff
@@ -172,7 +176,23 @@ impl Client {
...
+                    if self.check_website(&https_url).await.is_success() {
+                        Status::Error(
+                            format!("Insecure URL when HTTPS is available: {}", &url),
```
Comment (mre) [inline src/client.rs:187]: Could we provide an improvement suggestion to the user like this?

```rust
format!("An alternative secure URL is available. Consider replacing {} with {}.", &url, &https_url),
```

**Comment by mre** [comment]
Thanks for tackling this. Added a few minor comments. Otherwise looks good. 👍

File: src/client.rs (hunk 2 above)
Comment (lebensterben) [reply src/client.rs:181]: which module?

File: src/client.rs (hunk 4 above)
Comment (lebensterben) [reply src/client.rs]: Modified as your suggestion.

File: src/client.rs (hunk 3 above)
Comment (lebensterben) [reply src/client.rs:184]: I cannot use `?` because it returns `Result<(), ()>`, where the error `()` cannot be turned into `StdError`.

I can use `unwrap` or `expect`, which carries extra codes. Or I can use `unreachable!`, but it doesn't allow optimization.

From https://doc.rust-lang.org/stable/std/hint/fn.unreachable_unchecked.html#safety
> Use this function only when you can prove that the code will never call it
which is exactly the case.

File: src/client.rs (hunk 2 above)
Comment (lebensterben) [reply src/client.rs:181]: Actually, this is not exactly `is_https_available`.
The logic is, if `https://*` works, certainly it's available. But sometimes for example it times out due to slow connection, while the website actually provide `https` connection.

So this check is very conservative, it only returns true when `https` connection works at that particular time.

Well, in future we may do some re-tries (which should be configurable by user), for certain cases like time out.

Comment (mre) [reply src/client.rs:181]: Oh, I was just thinking of this section:

```rust
 let mut https_url = url.clone();
  https_url
      .set_scheme("https")
      .unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() });
  if self.check_website(&https_url).await.is_success() {
      Status::Error(
          format!(
              "An alternative secure URL is available. Consider replacing {} with {}.",
              &url, &https_url
          ),
          None,
      )
  } else {
      Status::Ok(code)
  }
```

What do you think about think about moving that into a function?

```rust
async fn is_https_available(url: &URL) -> bool {
   // ...
}
```

File: src/client.rs (hunk 3 above)
Comment (mre) [reply src/client.rs:184]: Aha! Didn't know that. 😄 Would still prefer an expect here for a few reasons:
* It documents our expectations
* avoids using `unsafe`, which could be a requirement for projects which want to use lychee as a library
* If this statement could ever fail (e.g. after refactoring), at least we get an error message that could help us fix it

Comment (lebensterben) [reply src/client.rs:184]: I can understand the first and third points. But the fear of `unsafe` is unreasonable...

Comment (lebensterben) [reply src/client.rs:184]: There's another approach
```rust
debug_assert!(foo.is_ok(), "some error message");
foo.unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() });
```

In debug build, you will get the same benefit of `expect`.
In release build, it allow optimization.

Comment (mre) [reply src/client.rs:184]: > I can understand the first and third points. But the fear of unsafe is unreasonable...

Hum, you might be right. At the same time my intuition is to use unsafe only if necessary. In this case, what would be the advantage of using unsafe? Is the perf/size/compile difference thanks to the optimizations  measurable? At runtime, we'll mostly be bound by I/O I'd say.

The other approach is clever, but it would take me a while to understand what's going on; I might not even get it without a comment.
So, in summary I'd like to keep it simple and use an `expect` here unless I'm overlooking a big drawback. 🤔

Comment (lebensterben) [reply src/client.rs:184]: >  what would be the advantage of using unsafe? Is the perf/size/compile difference thanks to the optimizations measurable?

Yes.

> At runtime, we'll mostly be bound by I/O I'd say.

That's also true.

> The other approach is clever, but it would take me a while to understand what's going on; I might not even get it without a comment.

```rust
debug_assert!(foo.is_ok(), "some error message");
foo.unwrap_or_else(|_| unsafe { std::hint::unreachable_unchecked() });
```

This is simple to understand. `debug_assert` is noop in release build, and in debug build it acts like `assert`.

By asserting the variable of interest is a `Ok`, it's safe to unwrap it anyway.
Also, `assert` allows you to add custom error message.

In conclusion, it well solved the two points you mentioned above.
> - It documents our expectations
> - If this statement could ever fail (e.g. after refactoring), at least we get an error message that could help us fix it

Again, there's just no solid reason why to avoid `unsafe` here. It's mathematically impossible to err.

Comment (mre) [reply src/client.rs:184]: Did some research. Performance gains are [debated](https://github.com/rust-lang/rust/issues/52945#issuecomment-409774203) in [this issue](https://github.com/rust-lang/rust/issues/52945) for the Rust compiler. It's arguably only faster in hot loops. We could do a benchmark to test that, but I'm not sure if it's worth it.

There's another comment by @oli-obk:

> I've been saved by the compiler panicking with "entered unreachable code" quite a few times. Especially during development where one might end up making such code reachable (maybe because a new features was implemented and some other code correctly assumed that it was unreachable before the existence of the new feature). Having these silently become UB seems... suboptimal.

He's a highly respected compiler dev, so I respect his opinion.
He mentions refactoring, which is also my argument. Even though the code is impossible to err, that might not be the case after refactoring. So, let's use expect.

Comment (lebensterben) [reply src/client.rs:184]: Nope, you're not listening.

The `debug_assert` test works even after refactoring.

Comment (lebensterben) [reply src/client.rs:184]: Put it in another way,

It an UB if it unwraps an Err. But the `debug_assert` explicitly check whether it's Err. If so, it panics with an error message, so that's what you can achieve with `expect`.

Then in release mode, it's OP, so no over head.

In one word, it's strictly better in every aspect than `expect`.

Comment (oli-obk) [reply src/client.rs:184]: Since I've been ~~summoned~~ pinged, I'm gonna throw in some unsolicited advice:

do benchmark the case using `unreachable_unchecked` against the unwrap case, and if you can really see a perf improvement here, do one of two things:

* debug asserts as suggested by @lebensterben 
* refactor the `set_scheme` function to return a `#[nonexhaustive]` empty enum (as the error type) instead of a `()` for now. You can add variants to that enum later if you decide that an error can actually happen. This way you get zero future API breakage (though you get it now), while the compiler can figure out that the `expect` call is dead code and eliminate it just as well as it would with the `unreachable_unchecked`.

If you don't see a perf improvement due to this `unreachable_unchecked`, either the compiler already figures it out for you, or this isn't really important. In either case my recommendation would be to stick with `expect`.

Comment (mre) [reply src/client.rs:184]: Thanks for the input @oli-obk. Makes a lot of sense. @lebensterben, would you be interested in working on the benchmark?

Comment (lebensterben) [reply src/client.rs:184]: Later. I'm struggling with the side effect of COVID vaccine. I'm in and out of sleep all day.

Comment (mre) [reply src/client.rs:184]: No rush. Hope you get well soon.

File: README.md [hunk 5]
```diff
@@ -164,6 +164,7 @@ FLAGS:
     -i, --insecure               Proceed for server connections considered insecure (invalid TLS)
     -n, --no-progress            Do not show progress bar. This is recommended for non-interactive shells (e.g. for
                                  continuous integration)
+        --prefer-https           When HTTPS is available, treat HTTP links as errors
```
Comment (pawroman) [inline README.md:167]: I think that some people will have quite different intuition for the semantics of "prefer HTTPS".

To me, the behavior introduced in this PR is more like "enforce HTTPS", e.g. "error out in case HTTPS is not available".

"Prefer HTTPS" would be something between the lines of "try HTTPS first, but if HTTP is the only option, then use that instead".

Maybe we should leave the prefer vs enforce part to the user? E.g. something like:

`--https-mode=enforce` and `--https-mode=prefer`?

When unspecified or specified as a default (`ignore`?) it would not change the current behavior. This way, we'd have maximum flexibility, and unambiguous semantics.

File: src/client.rs (hunk 2 above)
Comment (pawroman) [reply src/client.rs:181]: Yes, the code is quite dense -- a function would benefit the code here :+1:

**Review by pawroman — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (lebensterben) [reply README.md:167]: I will make a new commit and name it `require-https`

**Comment by lebensterben** [comment]
I rebased the code.
I renamed it to `--require-https`. (Forgotten to change commit message)
And replaced unsafe code.

But I didn't extract it to a separate function.

I just cannot come up with a meaningful name for this section. And I think this is quite straightforward already.

**Comment by pawroman** [comment]
> I renamed it to `--require-https`. (Forgotten to change commit message)
> And replaced unsafe code.

Thanks, that's much more intuitive now. Also cheers for removing unsafe!

> I just cannot come up with a meaningful name for this section. And I think this is quite straightforward already.

That might be the case, but still, the code is getting quite dense, and harder to reason about.

I think it would make sense to handle the entire `self.require_https` portion in a separate function, for instance, another branch of the if/else chain here: https://github.com/lycheeverse/lychee/pull/195/files#diff-c67c22c2f52695b3cdf26f7e6ba2fab2f49e4605941a7d87a005653d1c568e7aR178-R195

Could be changed to something like:
```rust
        let status = if self.filter.is_excluded(&uri) {
            Status::Excluded
        } else if uri.is_mail() {
            self.check_mail(&uri).await
        } else if self.require_https {
            // check website requiring HTTPS
            // (possibly a new method call, which would end up calling self.check_website)
        } else {
            self.check_website(&uri).await 
        };
```
This is because the logic is no longer "just return the status".

**Comment by lebensterben** [comment]
Well, the logic is different only when the URL intially returns Ok. Then we need to check whether it's HTTP and whether HTTPS is required. If so, change its scheme to HTTPS and check again. Should it returns Ok this time, we throw an error.

If create a separate function, it inevitably checks whether HTTPS is required for every URL. then in the body of new function, it either calls check_website twice, or repeat that parts of code. That's to say, it either created unnecessary function calls, or it increases redundancy.

**Comment by pawroman** [comment]
> That's to say, it either created unnecessary function calls, or it increases redundancy.

I don't think that's an issue. The compiler will likely optimize the code anyway. And even if there were no optimizations at all -- I believe the readability is worth it. Handling the "require HTTPS" case in the `else` branch simply obscures the "normal" case where we simply want to execute the request without any additional logic, specific to the "require HTTPS" case.

Moving it into a dedicated function would make it much more readable, easier to reason about (the reader would go "aha! the normal checking is not the same as checking when HTTPS is required"), and remove a few levels of code nesting. Also, any changes to the "require HTTPS" case would become more "localized" -- e.g. you'd only touch one function, without thinking much about it.

But that's, of course, all a bit subjective. I don't like nested if statements, as they make "executing the code in your mind" much harder, exponentially harder with every level of nesting.

**Comment by mre** [comment]
I agree with @pawroman here. Then again it's not really a blocker.
Added two tests and the pipeline is green, so I'm gonna merge this.

**Comment by MichaIng** [comment]
Great work guys! I just recently recognised a bunch of unnecessary plain HTTP links on our website, mostly leading to forced HTTPS redirects, but still, can be seen as unresponsable nowadays to lead visitors onto a non-encrypted track 🙂.
